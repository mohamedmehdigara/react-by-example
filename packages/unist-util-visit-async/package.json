{
  "name": "@app/unist-util-visit-async",
  "version": "1.0.0",
  "private": true,
  "types": "dist",
  "scripts": {
    "start": "nodemon -x \"run-s build\"",
    "build": "run-s build:*",
    "build:clean": "rimraf dist",
    "build:tsc": "tsc -p ./tsconfig.json",
    "test": "jest",
    "test:watch": "jest --watch"
  },
  "devDependencies": {
    "@app/examples": "workspace:*",
    "@app/templates": "workspace:*",
    "@types/jest": "26.0.13",
    "@types/unist": "2.0.3",
    "jest": "26.4.2",
    "nodemon": "2.0.4",
    "npm-run-all": "4.1.5",
    "remark-parse": "8.0.3",
    "remark-stringify": "8.1.1",
    "rimraf": "3.0.2",
    "ts-jest": "26.3.0",
    "typescript": "4.0.2",
    "unified": "9.2.0"
  }
}
