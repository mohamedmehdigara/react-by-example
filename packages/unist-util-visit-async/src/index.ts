import { Node, Parent } from 'unist'

const hasChildren = (node: Node): node is Parent => !!node.children

const visitWithParents = async (
  node: Node,
  visitor: (node: Node, parents: Parent[]) => void | Promise<void>,
  parents: Parent[],
): Promise<void> => {
  await visitor(node, parents)
  if (!hasChildren(node)) return
  const forEachChild = async (child: Node) => {
    await visitWithParents(child, visitor, [...parents, node])
  }
  await Promise.all(node.children.map(forEachChild))
}

export const visit = (
  root: Node,
  visitor: (node: Node, parents: Parent[]) => void | Promise<void>,
): Promise<void> => {
  return visitWithParents(root, visitor, [])
}
