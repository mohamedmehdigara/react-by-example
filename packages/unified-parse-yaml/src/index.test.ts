import { visit } from '@app/unist-util-visit-async'
import remarkParse from 'remark-parse'
import remarkStringify from 'remark-stringify'
import unified, { Processor } from 'unified'
import unifiedParseYaml from '.'

const fixtures = {
  simple: `\
# Hello world!

\`\`\`yaml {parse}
is: parsed
\`\`\`

\`\`\`yaml
is-not: parsed
\`\`\`

\`\`\`js
log:
console.log(\`
- hello
- this
- is
- a
- list
\`)
\`\`\`
`,
}

describe('unifiedParseYaml', () => {
  let processor: Processor

  describe('no filter', () => {
    beforeEach(() => {
      processor = unified()
        .use(remarkParse)
        .use(unifiedParseYaml)
        .use(remarkStringify)
        .use({ settings: { position: false } })
    })

    it.each(Object.entries(fixtures))(
      'should parse all values "%s"',
      async (_name, markdown) => {
        const ast = processor.parse(markdown)
        await processor.run(ast)
        const snapshot: [string, unknown][] = []
        await visit(ast, (node) => {
          if (node.data) snapshot.push([processor.stringify(node), node.data])
        })
        expect(snapshot).toMatchSnapshot()
      },
    )
  })

  describe('with filter', () => {
    beforeEach(() => {
      processor = unified()
        .use(remarkParse)
        .use(unifiedParseYaml, {
          filter: (n) =>
            n.type === 'code' && n.lang === 'yaml' && n.meta === '{parse}',
        })
        .use(remarkStringify)
        .use({ settings: { position: false } })
    })

    it.each(Object.entries(fixtures))(
      'should parse filtered values "%s"',
      async (_name, markdown) => {
        const ast = processor.parse(markdown)
        await processor.run(ast)
        const snapshot: [string, unknown][] = []
        await visit(ast, (node) => {
          if (node.data) snapshot.push([processor.stringify(node), node.data])
        })
        expect(snapshot).toMatchSnapshot()
      },
    )
  })
})
