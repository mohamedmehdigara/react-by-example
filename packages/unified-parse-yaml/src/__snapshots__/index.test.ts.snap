// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unifiedParseYaml no filter should parse all values "simple" 1`] = `
Array [
  Array [
    "\`\`\`yaml {parse}
is: parsed
\`\`\`",
    Object {
      "yaml": Object {
        "is": "parsed",
      },
    },
  ],
  Array [
    "\`\`\`yaml
is-not: parsed
\`\`\`",
    Object {
      "yaml": Object {
        "is-not": "parsed",
      },
    },
  ],
  Array [
    "Hello world!",
    Object {
      "yaml": "Hello world!",
    },
  ],
]
`;

exports[`unifiedParseYaml with filter should parse filtered values "simple" 1`] = `
Array [
  Array [
    "\`\`\`yaml {parse}
is: parsed
\`\`\`",
    Object {
      "yaml": Object {
        "is": "parsed",
      },
    },
  ],
]
`;
